src/Transport/Receiving/MessagePump.cs

```diff
@@ -0,0 +1,295 @@
+        MessageReceiver receiver;
+
+        static readonly ILog logger = LogManager.GetLogger<MessagePump>();
+
+        public MessagePump(ServiceBusConnectionStringBuilder connectionStringBuilder, ITokenProvider tokenProvider, int prefetchMultiplier, int overriddenPrefetchCount, TimeSpan timeToWaitBeforeTriggeringCircuitBreaker)
+        {
+            this.connectionStringBuilder = connectionStringBuilder;
+            this.tokenProvider = tokenProvider;
+            this.prefetchMultiplier = prefetchMultiplier;
+            this.overriddenPrefetchCount = overriddenPrefetchCount;
+            this.timeToWaitBeforeTriggeringCircuitBreaker = timeToWaitBeforeTriggeringCircuitBreaker;
+        }
+
+        public Task Init(Func<MessageContext, Task> onMessage, Func<ErrorContext, Task<ErrorHandleResult>> onError, CriticalError criticalError, PushSettings settings)
+        {
+            if (settings.PurgeOnStartup)
+            {
+                throw new Exception("Azure Service Bus transport doesn't support PurgeOnStartup behavior");
+            }
+
+            this.onMessage = onMessage;
+            this.onError = onError;
+            pushSettings = settings;
+
+            circuitBreaker = new RepeatedFailuresOverTimeCircuitBreaker($"'{settings.InputQueue}'", timeToWaitBeforeTriggeringCircuitBreaker, criticalError);
+
+            return Task.CompletedTask;
+        }
+
+        public void Start(PushRuntimeSettings limitations)
+        {
+            maxConcurrency = limitations.MaxConcurrency;
+
+            var prefetchCount = overriddenPrefetchCount;
+
+            if (prefetchCount == 0)
+            {
+                prefetchCount = maxConcurrency * prefetchMultiplier;
+            }
+
+            var receiveMode = pushSettings.RequiredTransactionMode == TransportTransactionMode.None ? ReceiveMode.ReceiveAndDelete : ReceiveMode.PeekLock;
+
+            if (tokenProvider == null)
+            {
+                receiver = new MessageReceiver(connectionStringBuilder.GetNamespaceConnectionString(), pushSettings.InputQueue, receiveMode, retryPolicy: default, prefetchCount);
+            }
+            else
+            {
+                receiver = new MessageReceiver(connectionStringBuilder.Endpoint, pushSettings.InputQueue, tokenProvider, connectionStringBuilder.TransportType, receiveMode, retryPolicy: default, prefetchCount);
+            }
+
+            semaphore = new SemaphoreSlim(maxConcurrency, maxConcurrency);
+
+            messageProcessing = new CancellationTokenSource();
+
+            receiveLoopTask = Task.Run(() => ReceiveLoop());
+        }
+
+        async Task ReceiveLoop()
+        {
+            try
+            {
+                while (!messageProcessing.IsCancellationRequested)
+                {
+                    await semaphore.WaitAsync(messageProcessing.Token).ConfigureAwait(false);
+
+                    var receiveTask = receiver.ReceiveAsync();
+
+                    ProcessMessage(receiveTask)
```

Comment:
Can we simplify this to use non-continuation style?

Reply:
IIRC we needed continuation style. @bording do you recall details?

Reply:
If we can get the same behavior without it, I don't think it's required. It was the best way we came up with to get the characteristics we wanted here, though.

Reply:
Raised issue #31